Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,194 @@
+        self._compile_program("fontProgram", "fpgm")
+
+    def compile_glyf(self):
+        for name in sorted(self.ufo.keys()):
+            glyph = self.ufo[name]
+            ttdata = glyph.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+            if name not in self.font["glyf"]:
+                if ttdata is not None:
+                    logger.warning(
+                        f"Glyph '{name}' not found in font, "
+                        "skipping compilation of TrueType instructions "
+                        "for this glyph."
+                    )
+                logger.debug(f"UFO keys: {list(self.ufo.keys())}")
+                logger.debug(f"glyf keys: {list(self.font['glyf'].keys())}")
+                continue
+
+            glyf = self.font["glyf"][name]
+            if ttdata is not None:
+                formatVersion = ttdata.get("formatVersion", None)
+                if int(formatVersion) != 1:
+                    logger.error(
+                        f"Unknown formatVersion {formatVersion} "
+                        "in glyph '{name}', it will have "
+                        "no instructions in font."
+                    )
+                    continue
+
+                # Check if glyph hash matches the current outlines
+                hash_pen = HashPointPen(glyph.width, self.ufo)
+                glyph.drawPoints(hash_pen)
+                glyph_id = ttdata.get("id", None)
+                if glyph_id is None or glyph_id != hash_pen.hash:
+                    logger.error(
+                        f"Glyph hash mismatch, glyph '{name}' will have "
+                        "no instructions in font."
+                    )
+                    continue
+
+                # Compile the glyph program
+                asm = ttdata.get("assembly", None)
+                if asm is not None:
+                    glyf.program = ttLib.tables.ttProgram.Program()
+                    glyf.program.fromAssembly(asm)
+                    # Roundtrip once, or if the font is dumped to XML before
+                    # having been saved, the assembly code if will look awful.
+                    glyf.program.fromBytecode(glyf.program.getBytecode())
+
+            # Handle composites
+            if glyf.isComposite():
+                # Remove empty glyph programs from composite glyphs
+                if hasattr(glyf, "program") and not glyf.program:
+                    delattr(glyf, "program")
+
+                # Set component flags
+
+                # We need to decide when to set the flags.
+                # Let's assume if any lib key is not there, or the component
+                # doesn't have an identifier, we should leave the flags alone.
+                use_my_metrics_comp = None
+                for i, c in enumerate(glyf.components):
+                    if i >= len(glyph.components):
+                        logger.error(
+                            "Number of components differ between UFO and TTF "
```

Comment:
that'd be weird if it happened, should it not be an error?

Reply:
It happens when compileTTF is called with `flattenComponents=True`. I'd prefer not to raise an error, but I've changed it so 1) it is tested before setting the component flags, and 2) it skips setting component flags for the glyph completely.